src/jvm/main/org/jetbrains/kotlinx/lincheck/Utils.kt

```diff
@@ -329,3 +329,20 @@ internal inline fun <R> runOutsideIgnoredSection(descriptor: ThreadDescriptor?,
 
 internal const val LINCHECK_PACKAGE_NAME = "org.jetbrains.kotlinx.lincheck."
 internal const val LINCHECK_RUNNER_PACKAGE_NAME = "org.jetbrains.kotlinx.lincheck.runner."
+
+internal fun <T> Class<T>.newInstanceRecursive(): T {
+    @Suppress("UNCHECKED_CAST")
+    val constructor = this.declaredConstructors.singleOrNull { it.parameterCount == 0 } as? Constructor<T>
+    if (constructor != null) {
+        return constructor.newInstance()
+    }
+
+    if (this.enclosingClass != null) {
```

Comment:
Shouldn't we check if `enclosingClass != null` as a first step, and only then look for default constructor with zero arguments?

Reply:
Not really. Class can have a default constructor even if `enclosingClass != null`. For example
```
package org.example

class A {
    inner class B
}

class C {
    class D
}

fun main() {
    println(A.B::class.java.enclosingClass) // class org.example.A
    println(C.D::class.java.enclosingClass) // class org.example.C
    // But
    A().B() // requires A object
    C.D()   // doesn't require C object
}
```

I can probably add such a test.

Reply:
Ok, I see, thanks. 

Yes, it would be nice to add another test case for nested non-inner class (static nested in Java terminology).